File: sdk/python/memg/decay.py
"""Decay and pruning: remove expired facts and stale summaries."""

from __future__ import annotations

import logging
from typing import Tuple

from .store import Store

logger = logging.getLogger("memg")


def prune_expired_and_stale(
    store: Store,
    summary_max_age_days: int = 90,
) -> Tuple[int, int]:
    """Prune expired facts and stale conversation summaries.

    Logic matches Go memory/decay.go:
    - Call store.prune_expired_facts with empty entity (all entities).
    - Call store.prune_stale_summaries.

    Returns (facts_pruned, summaries_cleared).
    """
    facts_pruned = 0
    try:
        facts_pruned = store.prune_expired_facts("")
    except Exception:
        logger.warning("memg pruner: prune expired facts failed", exc_info=True)

    summaries_cleared = 0
    try:
        summaries_cleared = store.prune_stale_summaries(summary_max_age_days)
    except Exception:
        logger.warning("memg pruner: prune stale summaries failed", exc_info=True)

    return facts_pruned, summaries_cleared

File: sdk/typescript/src/decay.ts
/**
 * Decay: prune expired facts and stale summaries.
 * Matches the Go memory/decay.go Pruner logic.
 */

import type { Store } from './store.js';

/**
 * Prune expired facts and stale conversation summaries.
 *
 * 1. Calls store.pruneExpiredFacts with current time (batched internally by store).
 * 2. Calls store.pruneStaleSummaries with the given max age (default 90 days).
 * 3. Returns counts of pruned facts and cleared summaries.
 */
export async function pruneExpiredAndStale(
  store: Store,
  summaryMaxAgeDays?: number
): Promise<{ factsPruned: number; summariesCleared: number }> {
  const now = new Date().toISOString();
  const maxAge = summaryMaxAgeDays ?? 90;

  // Prune expired facts. The store implementation batches internally.
  // Pass empty entityUuid to prune across all entities (matching Go behavior
  // which passes "" for the global sweep).
  const factsPruned = await store.pruneExpiredFacts('', now);
  const summariesCleared = await store.pruneStaleSummaries(maxAge);

  return { factsPruned, summariesCleared };
}
